### DISCUSSION TITLE
How can I "test" in Spring AI?

### DISCUSSION THREAD

**Discussion by devholic22** [spring-projects/spring-ai#679] [Q&A] [answered]
Hello, I'm a student who love Spring and Java.

when I use RestTemplate or WebClient for function with AI (ex: call ChatGPT to ask concept of something), I can test it
with JUnit, Mockito. (create mock function)

but I don't know how I can test AI function when I use Spring AI.

thank you for read my question.

**Comment by f3rnandomoreno** [comment, marked as answer]
You have many examples in the project of the integrationTests. They are files finished in IT, for example this one

https://github.com/spring-projects/spring-ai/blob/1c93ae50a805dcb54b05a6fbbc11667b3aac6562/models/spring-ai-ollama/src/test/java/org/springframework/ai/ollama/OllamaChatClientIT.java

You can find also for OpenAi chat client and many more. I think they can guide you.

**Reply by devholic22** [reply]
Thank you for reply my question.

When I thought about it, I thought I could make enough mock objects through the interfaces in Spring AI to do unit tests.

The integrated test you mentioned will also be a necessary process for the stability of the system, but in the actual situation, tokens will be consumed every time we do an integrated test, so I would like to ask if this is a reasonable judgment in terms of cost.

what do you think about this question?

**Reply by devholic22** [reply]
I saw `OpenAiChatClientWithChatResponseMetadataTests` and found out about the `MockRestServiceServer` that can mock external API calls themselves. Thank you!
